runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaMetricOp.java

```diff
@@ -0,0 +1,102 @@
+ * Runner. A MetricOp can be either attached to Input PCollection or Output PCollection of a
+ * PTransform.
+ *
+ * <p>A MetricOp is created per primitive PTransform per PCollection its across its inputs &
+ * outputs. 1. An independent MetricOp is created and attached to each input PCollection to the
+ * PTransform. 2. An independent MetricOp is created and attached to each input PCollection to the
+ * PTransform.
+ *
+ * <p>Each concrete MetricOp is responsible for following metrics computation: 1. Throughput: Emit
+ * the number of elements processed in the PCollection 2. Watermark Progress: Emit the watermark
+ * progress of the PCollection 3. Latency: Maintain the avg arrival time per watermark across
+ * elements it processes, compute & emit the latency
+ *
+ * @param <T> type of the message
+ */
+public abstract class SamzaMetricOp<T> implements Op<T, T, Void> {
+  // Unique name of the PTransform this MetricOp is associated with
+  protected final String transformFullName;
+  protected final BeamTransformMetricRegistry beamTransformMetricRegistry;
+  // Name or identifier of the PCollection which Ptrasform is processing
+  protected final String pValue;
+  // List of input PValue(s) for all PCollections processing the PTransform
+  protected List<String> transformInputs;
+  // List of output PValue(s) for all PCollections processing the PTransform
+  protected List<String> transformOutputs;
+  // Name of the task, for logging purpose
+  protected String task;
+
+  // Some fields are initialized in open() method, which is called after the constructor.
+  @SuppressWarnings("initialization.fields.uninitialized")
+  public SamzaMetricOp(
+      String pValue,
+      String transformFullName,
+      BeamTransformMetricRegistry beamTransformMetricRegistry) {
+    this.transformFullName = transformFullName;
+    this.beamTransformMetricRegistry = beamTransformMetricRegistry;
+    this.pValue = pValue;
+  }
+
+  @Override
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  public void open(
+      Config config,
+      Context context,
+      Scheduler<KeyedTimerData<Void>> timerRegistry,
+      OpEmitter<T> emitter) {
+    final Map.Entry<String, String> transformInputOutput =
+        PipelineJsonRenderer.getTransformIOMap(config).get(transformFullName);
+    this.transformInputs =
+        transformInputOutput != null
+            ? ioFunc(transformInputOutput.getKey()).get()
+            : new ArrayList();
+    this.transformOutputs =
+        transformInputOutput != null
+            ? ioFunc(transformInputOutput.getValue()).get()
+            : new ArrayList();
+    // for logging / debugging purposes
+    this.task = context.getTaskContext().getTaskModel().getTaskName().getTaskName();
+    // Register the transform with BeamTransformMetricRegistry
+    beamTransformMetricRegistry.register(transformFullName, pValue, context);
+  }
+
+  private static Supplier<List<String>> ioFunc(String ioList) {
```

Comment:
Why not just use a normal `private static List<String> ioFunc(String ioList)`, instead of the Supplier?

Reply:
see line 88 and 84

Reply:
We can write line 88 and 84 as `ioFunc(transformInputOutput.getValue())`
and 

```
private static List<String> ioFunc(String ioList) {
  return Arrays.stream(ioList.split(PipelineJsonRenderer.TRANSFORM_IO_MAP_DELIMITER))
            .filter(item -> !item.isEmpty())
            .collect(Collectors.toList());
}
```

why do we need async here?

Reply:
o i see yup not needed